content/advent-2017/mphf.md

```diff
@@ -0,0 +1,297 @@
+	for i := 0; i < b.N; i++ {
+		for _, k := range keys32 {
+			sink = values[hash2(k)] == k
+		}
+	}
+}
+
+func BenchmarkMap(b *testing.B) {
+	m := make(map[uint32]bool)
+	for _, k := range keys32 {
+		m[k] = true
+	}
+
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		for _, k := range keys32 {
+			sink = m[k]
+		}
+	}
+}
+```
+
+And the output:
+
+```
+BenchmarkHash-8   	100000000	        20.4 ns/op
+BenchmarkMap-8    	10000000	       172 ns/op
+```
+
+A very significant improvement.
+
+But how can we generalize this? And does this always work? And is it always
+fast?
+
+Well, the first thing we notice is that as the set becomes larger, it becomes
+much more difficult to find a value for `multiplier` that works, and one might
+not even exist. Even with 32 elements, trying to find the right
+value for `multiplier` might be cost prohibitive or even impossible. We
+certainly wouldn’t want to do it at runtime. These questions are why Minimal
+Perfect Hash Functions are an interesting research topic.
+
+*Hash-Displace*
+
+The issue we ran into with our second attempt was that it was too difficult to
+find a single value for the multiplier that worked for larger sets. Two keys
+that collide with one hash function are unlikely to collide with a second hash
+function as well. We can take advantage of this fact by switching to a
+construction that uses more than one hash function. The first hash function
+will distribute the keys over the buckets, with "a few" collisions.  Then, for
+each set of collisions, we try to find a second hash function that distributes
+the keys evenly with no collisions. The paper fully describing this approach is
+[Hash, Displace, and Compress](http://cmph.sourceforge.net/papers/esa09.pdf)
+and, an earlier version is [Practical Minimal Perfect Hashing Functions for
+Large Databases](http://eprints.cs.vt.edu/archive/00000223/01/TR-90-41.pdf).
+My simplified version of this algorithm is here:
+https://github.com/dgryski/go-mph .
+
+My implementation is about 25% faster than using a regular Go map for 16 keys
+and 50% faster when testing with the 235,000 entries in */usr/dict/words*.
+Unlike the previous algorithm, this one has no issues with large key sets.
+Constructing the hash function for this wordlist takes only 100ms-125ms.
+
+There are three ways to judge a hash function: construction time, evaluation
+time, and space usage. Our first successful hash function had virtually no
+space usage, a fast evaluation time, but a huge construction time.
+
+This algorithm has a pretty small construction cost. It turns out to be
+linear in the number of keys. The evaluation time is also constant time: one
+"standard" hash function evaluation, some integer mixing, and two table
+lookups.
+
+As for space usage, in my implementation this algorithm uses 8 bytes per
+entry: each key gets its own 4-byte index (0..N-1) and another 4-byte seed
+for the second hash function. The paper gives a method that allows the
```

Comment:
By `paper`, you mean `Hash, Displace, and Compress`?

Reply:
Yes.  I'll specify that.